### PR TITLE
Prometheus exports for lnd

### PR DESCRIPTION
**Pull request by baryluk**
This implements base of exports for monitoring.

Server, channels, peers, invoices, transactions, grpc, and wire messages are being monitored.

In total about 1000 counters is being tracked (some in vector metrics, some automatically from Go runtime and http and gRPC servers), across 80 different metrics.

```text
$ curl -s "http://[::1]:8081/metrics" | egrep -i '^# HELP'
# HELP go_gc_duration_seconds A summary of the GC invocation durations.
# HELP go_goroutines Number of goroutines that currently exist.
# HELP go_info Information about the Go environment.
# HELP go_memstats_alloc_bytes Number of bytes allocated and still in use.
# HELP go_memstats_alloc_bytes_total Total number of bytes allocated, even if freed.
# HELP go_memstats_buck_hash_sys_bytes Number of bytes used by the profiling bucket hash table.
# HELP go_memstats_frees_total Total number of frees.
# HELP go_memstats_gc_cpu_fraction The fraction of this program's available CPU time used by the GC since the program started.
# HELP go_memstats_gc_sys_bytes Number of bytes used for garbage collection system metadata.
# HELP go_memstats_heap_alloc_bytes Number of heap bytes allocated and still in use.
# HELP go_memstats_heap_idle_bytes Number of heap bytes waiting to be used.
# HELP go_memstats_heap_inuse_bytes Number of heap bytes that are in use.
# HELP go_memstats_heap_objects Number of allocated objects.
# HELP go_memstats_heap_released_bytes Number of heap bytes released to OS.
# HELP go_memstats_heap_sys_bytes Number of heap bytes obtained from system.
# HELP go_memstats_last_gc_time_seconds Number of seconds since 1970 of last garbage collection.
# HELP go_memstats_lookups_total Total number of pointer lookups.
# HELP go_memstats_mallocs_total Total number of mallocs.
# HELP go_memstats_mcache_inuse_bytes Number of bytes in use by mcache structures.
# HELP go_memstats_mcache_sys_bytes Number of bytes used for mcache structures obtained from system.
# HELP go_memstats_mspan_inuse_bytes Number of bytes in use by mspan structures.
# HELP go_memstats_mspan_sys_bytes Number of bytes used for mspan structures obtained from system.
# HELP go_memstats_next_gc_bytes Number of heap bytes when next garbage collection will take place.
# HELP go_memstats_other_sys_bytes Number of bytes used for other system allocations.
# HELP go_memstats_stack_inuse_bytes Number of bytes in use by the stack allocator.
# HELP go_memstats_stack_sys_bytes Number of bytes obtained from system for stack allocator.
# HELP go_memstats_sys_bytes Number of bytes obtained from system.
# HELP go_threads Number of OS threads created.
# HELP grpc_server_handled_total Total number of RPCs completed on the server, regardless of success or failure.
# HELP grpc_server_msg_received_total Total number of RPC stream messages received on the server.
# HELP grpc_server_msg_sent_total Total number of gRPC stream messages sent by the server.
# HELP grpc_server_started_total Total number of RPCs started on the server.
# HELP lnd_announcement_batch_count Number of processed (broadcasted) batches with announcements.
# HELP lnd_announcement_batch_size_total Total number of all announcements in all processed (broadcasted) batches.
# HELP lnd_best_header_timestamp What is the header timestamp of the best synced block in a chain.
# HELP lnd_block_height Height of the best chain.
# HELP lnd_chain_transaction_count Number of all transactions on chain.
# HELP lnd_chain_transaction_fee_total Total of all fees across all transactions on chain.
# HELP lnd_chain_transaction_received_amount_total Total amount of coins in all transactions with positive amounts.
# HELP lnd_chain_transaction_sent_amount_total Total amount of coins in all transactions with negative amounts.
# HELP lnd_channel_balance_total Total .
# HELP lnd_channel_capacity_by_channel L...
# HELP lnd_channel_commit_weight_by_channel L...
# HELP lnd_channel_csv_delay_by_channel Local channel config CSV delay
# HELP lnd_channel_external_commit_fee_by_channel L...
# HELP lnd_channel_fee_per_kw_by_channel L...
# HELP lnd_channel_local_balance_by_channel Local balance of a channel by channel
# HELP lnd_channel_pending_htlcs_count_by_channel Local commit HTLCs count
# HELP lnd_channel_pending_open_balance_total Total .
# HELP lnd_channel_remote_balance_by_channel Remote balance of a channel by channel
# HELP lnd_channel_satoshis_received_by_channel L...
# HELP lnd_channel_satoshis_sent_by_channel L...
# HELP lnd_channel_updates_count_by_channel Local commit height
# HELP lnd_channels_active_count Number of active channels across all peers.
# HELP lnd_channels_pending_count Number of pending channels.
# HELP lnd_graph_edge_count What is the number of known edges in the network.
# HELP lnd_graph_node_count What is the size of know network.
# HELP lnd_invoices_count Number of all invoices.
# HELP lnd_invoices_pending_count Number of pending invoices.
# HELP lnd_new_channels_discovered_count Number of new channels discovered.
# HELP lnd_peer_bytes_recv_by_peer Bytes sent to peer by peer.
# HELP lnd_peer_bytes_sent_by_peer Satosish sent to peer by peer.
# HELP lnd_peer_count Number of server peers.
# HELP lnd_peer_ping_by_peer Ping to peer by peer.
# HELP lnd_peer_sat_recv_by_peer Satosish sent to peer by peer.
# HELP lnd_peer_sat_sent_by_peer Satosish sent to peer by peer.
# HELP lnd_synced_to_chain Is LND synced?
# HELP lnd_version Version of LND running.
# HELP lnd_wallet_confirmed_balance Total balance, from txs that have >= 1 confirmations.
# HELP lnd_wallet_total_balance Total balance, from txs that have >= 0 confirmations.
# HELP lnd_wire_message_received_by_type_count Total number of wire messages received by type.
# HELP lnd_wire_message_sent_by_type_count Total number of wire messages (to be sent) by type.
# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.
# HELP process_max_fds Maximum number of open file descriptors.
# HELP process_open_fds Number of open file descriptors.
# HELP process_resident_memory_bytes Resident memory size in bytes.
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# HELP process_virtual_memory_bytes Virtual memory size in bytes.
# HELP promhttp_metric_handler_requests_in_flight Current number of scrapes being served.
# HELP promhttp_metric_handler_requests_total Total number of scrapes by HTTP status code.
```

Closes: https://github.com/lightningnetwork/lnd/issues/1243

### CODE REVIEW

File: peer.go [hunk 1]
```diff
@@ -1236,6 +1240,94 @@ func (p *peer) logWireMessage(msg lnwire.Message, read bool) {
 	}))
 }
 
+// messageTypeString returns a human and monitoring-readable string that describes an
+// incoming/outgoing message type.
+func messageTypeString(msg lnwire.Message) string {
```
Comment (Roasbeef) [inline peer.go]: The `String` method of `lnwire.Message` already does this. My guess is that this is in place due to naming conventions that prefer snake case for metrics? In that case, we method can easily be made to convert the camel case names to snake case. The bonus of this is that then we wouldn't need to maintain the mapping in two distinct areas within the codebase.

Comment (baryluk) [reply peer.go]: No, camel case is also fine. I didn't realise it is a protobuf message with String method. I will make the code use the String method!

Comment (Roasbeef) [reply peer.go]: Here's the method I'm talking about: https://github.com/lightningnetwork/lnd/blob/master/lnwire/message.go#L59

**Comment by baryluk** [comment]
As of dependencies, AFAIK `lnd` already does use parts of github.com/grpc-ecosystem. Also the dependencies:

`github.com/grpc-ecosystem/go-grpc-middleware`
`github.com/grpc-ecosystem/go-grpc-prometheus`
`github.com/prometheus/client_golang`

are rather thin.

`github.com/prometheus/common` - This appears to be to allow either text or protocol buffer output from "/metrics". It also provides json output format, instead of simple text format (and optionally removes the HELP and TYPE fields probably too). From benchmarks it looks like binary protobuf output is almost 6 times faster than text, so I can see why this is useful. This package also handle things like escaping of characters that could could interfere with label and variable names in text format. I was able to trigger this by `curl -H "Accept: application/vnd.google.protobuf; proto=io.prometheus.client.MetricFamily; encoding=compact-text" -s "http://[::1]:8081/metrics" `. I belive the prometheus automatically adds all relevant Accept types in HTTP headers, and client automatically selects best one (and sets Content-Type header). That is nice.
`github.com/prometheus/client_model` - It only contains single go file, which looks to be generated from protobuf, and is used by `common` above for transport of metrics and some enums for value types.
`github.com/prometheus/procfs` - obviously for exporting various process information.

The only ones that I would be concerned are these:

`github.com/matttproud/golang_protobuf_extensions`

This is not about protobuf extensions, but rather extensions to the Go API. It adds `pbutil.ReadDelimited` and `pbutil.WriteDelimited`. At the momement I do not know what these are used. These were proposed to be added to official Go protobuf package, https://codereview.appspot.com/9102043/ , but it was rejected apparently. The code is very similar, it is just a simple utility to storing multiple proto messages.

`github.com/beorn7/perks` - this is for histogram and summary computations (i.e. quantiles and averages over time), in go client library, but at the moment they are not used. I do plan to add latency histograms tho. The entire package is just 2 files.

These are very small, but are maintained by single person in their personal repo, so that is a bit worrying, unless we add overrides and pin specific commits.

Without this PR:

```text
$ ls -l lnd
-rwxr-xr-x 1 baryluk baryluk 31551006 Jun 11 02:20 lnd
$ size lnd
   text	   data	    bss	    dec	    hex	filename
18427844	 330705	 156968	18915517	120a0bd	lnd
```

With this PR:

```text
$ ls -l lnd
-rwxr-xr-x 1 baryluk baryluk 32914066 Jun 11 02:14 lnd
$ size lnd
   text	   data	    bss	    dec	    hex	filename
19123104	 331697	 157480	19612281	12b4279	lnd
$
```

Which is +1.3MB in file size. and 700KB in text size. I am building using `go build`, so maybe this is just a debug data?

With `go build` + strip, I got file sizes:

Without this PR:

```
-rwxr-xr-x 1 baryluk baryluk 18768736 Jun 11 02:22 lnd
```

With this PR:

```
-rwxr-xr-x 1 baryluk baryluk 19466048 Jun 11 02:24 lnd
```

A difference of 700KB, as expected.

I think a lot of this is actually never touched during execution. This is a bit more than I expected, and I can try and hunt what is the main contributor. I.e. I can remove the 


As far as performance goes. There is no CPU impact if one does not use it at all (increment of some counters, i.e. in peer, or in grpc, is extremely fast). And memory impact is very small. Less than 50KB.

**Comment by vapopov** [comment]
@baryluk Hi! Could you please share somewhere dashboard file for grafana. It's very useful implementation and I would like to use it already for our instances :)

**Comment by baryluk** [comment]
> @baryluk Hi! Could you please share somewhere dashboard file for grafana. It's very useful implementation and I would like to use it already for our instances :)

In attachment, the job is hardcoded to "lnd" in few graphs, otherwise Go related runtime statistics will pull out information runtime of Prometheus itself too and any other Go jobs monitored. I am not familiar with Grafana to know how to properly parameterize the dashboard to make it easier. Hopefully it is useful and easy to fix manually.

[LND monitoring-1543090963379.json.gz](https://github.com/lightningnetwork/lnd/files/2612577/LND.monitoring-1543090963379.json.gz)

**Comment by baryluk** [comment]
ping.

I am not going to maintain this branch forever.

**Comment by Roasbeef** [comment]
Superseded by #3134 (went with a minimal approach), follow up projects coming soon.

**Pull request closed without merging**